### PR TITLE
core: optimize mergeDistanceRangeMaps (use putMany)

### LINKED ISSUE #5843: core: use optimized putMany to improve DistanceRangeMap performance
### Description and goal

See https://github.com/osrd-project/osrd/issues/5513
We should update calls and measure impact

### Acceptance criteria

TBD

### PR DESCRIPTION
**Pull request by shenriotpro**
fixes https://github.com/osrd-project/osrd/issues/5843

### CODE REVIEW

File: core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMap.kt [hunk 1]
```diff
@@ -61,3 +61,23 @@ fun <T> distanceRangeMapOf(entries: List<DistanceRangeMap.RangeMapEntry<T>> = em
   : DistanceRangeMap<T> {
     return DistanceRangeMapImpl(entries)
 }
+
+/** Merges all the given range maps, offsetting them by the given distances.
+ * The lists must be empty or `maps` must be larger by one. */
+fun <T> mergeDistanceRangeMaps(maps: List<DistanceRangeMap<T>>, distances: List<Distance>): DistanceRangeMap<T> {
+        assert((maps.size - 1 == distances.size) || (maps.isEmpty() && distances.isEmpty()))
+
+        val res_entries = ArrayList<DistanceRangeMap.RangeMapEntry<T>>()
```
Comment (eckter) [inline core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMap.kt]:
```suggestion
        val resEntries = ArrayList<DistanceRangeMap.RangeMapEntry<T>>()
```

File: core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMap.kt [hunk 2]
```diff
@@ -61,3 +61,23 @@ fun <T> distanceRangeMapOf(entries: List<DistanceRangeMap.RangeMapEntry<T>> = em
...
+        for ((map, distance) in maps zip distances + 0.meters) {
+            for (entry in map) {
+                res_entries.add(DistanceRangeMap.RangeMapEntry(entry.lower + previousDistance, entry.upper + previousDistance, entry.value))
+            }
+            previousDistance += distance
+        }
+
+        // Build the whole map at once to avoid redundant computations.
+        return distanceRangeMapOf(res_entries)
+    }
```
Comment (eckter) [inline core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/DistanceRangeMap.kt:79]: The function has one extra level of indentation

File: core/kt-osrd-utils/src/test/kotlin/fr/sncf/osrd/utils/TestDistanceRangeMap.kt [hunk 3]
```diff
@@ -221,4 +221,53 @@ class TestDistanceRangeMap {
...
+    }
+
+    @Test
+    fun testMergeDistanceRangeMapsLarge() {
+        val n = 200
+        val oneSecond: Duration = 1.seconds
+        val timeSource = TimeSource.Monotonic
+        val distances = List(n-1) { Distance(n.toLong()) }
+        var maps: MutableList<DistanceRangeMap<Int>> = mutableListOf()
+        for (i in 0..n-1) {
```
Comment (eckter) [inline core/kt-osrd-utils/src/test/kotlin/fr/sncf/osrd/utils/TestDistanceRangeMap.kt]:
```suggestion
        for (i in 0..<n) {
```

File: core/kt-osrd-utils/src/test/kotlin/fr/sncf/osrd/utils/TestDistanceRangeMap.kt [hunk 4]
```diff
@@ -221,4 +221,53 @@ class TestDistanceRangeMap {
...
+            ), rangeMap.asList())
+    }
+
+    @Test
+    fun testMergeDistanceRangeMapsLarge() {
+        val n = 200
+        val oneSecond: Duration = 1.seconds
+        val timeSource = TimeSource.Monotonic
+        val distances = List(n-1) { Distance(n.toLong()) }
+        var maps: MutableList<DistanceRangeMap<Int>> = mutableListOf()
```
Comment (eckter) [inline core/kt-osrd-utils/src/test/kotlin/fr/sncf/osrd/utils/TestDistanceRangeMap.kt]:
```suggestion
        val maps: MutableList<DistanceRangeMap<Int>> = mutableListOf()
```

**Review by eckter — APPROVED** [review]
A few nitpicky comments on the formatting, but other than that LGTM

**Review by Erashin — APPROVED** [review]

**Pull request merged**